\documentclass{article}
\usepackage{tikz}
\usepackage{array}
\usepackage{graphicx}
\usepackage{lscape}
\usepackage{amsmath}
\usepackage{multirow}
\usepackage{hyperref}

\DeclareMathOperator{\SJ}{SJ}

\begin{document}
\title{Fast quantification of splice junctions by {\em sjcount}}
\author{Dmitri D. Pervouchine}
\date{\today}
\maketitle

\section{Synopsis}

The purpose of {\em sjcount} is to provide a fast method for quantification of splice junctions from BAM files. It is an annotation-agnostic version of 
\href{https://github.com/pervouchine/bam2ssj}{bam2ssj}.


\section{Installation and usage}

See README.md file for installation instructions. {\em sjcount} is used with the following keys

\begin{verbatim}
sjcount -bam bam_file [-ssj junctions_output] [-ssc boundary_output]
       [-maxlen max_intron_length] [-minlen min_intron_length] [-margin length] 
       [-read1 0|1] [-read2 0|1] [-nbins number_of_bins] [-binsize bin_size] 
       [-lim number_of_lines] [-quiet]
\end{verbatim}
where
\begin{itemize}
\item {\bf bam\_file} is a sorted input BAM file with a header
\item {\bf junctions\_output} is the output file with junction counts
\item {\bf boundary\_output} is the output file with boundary counts
\item {\bf maxlen} upper limit on intron length, 0 = no limit (default=0)
\item {\bf minlen} lower limit on intron length, 0 = no limit (default=0)
\item {\bf margin} length, see below, (default=0)
\item {\bf read1} 0/1, reverse complement read1 no/yes (default=no)
\item {\bf read2} 0/1, reverse complement read2 no/yes (default=no)
\item {\bf binsize} size of the overhang bin, (default=$\infty$)
\item {\bf nbins} number of overhang bins, (default=1)
\item {\bf lim} nreads stop after nreads, (default=no limit)
\item {\bf quiet} -- suppress verbose output
\end{itemize}

The output consists of two parts. First, a tab-delimited file containing splice junction counts is produced.
Its format is as follows
\begin{verbatim}
chr1    100	200	-1	10	25
chr1    100     200     -1      11      12
...	...	...	...	...	...
\end{verbatim}
where the first column contains chromosome id, the second and the third columns contain positions of terminal exonic 
nucleotides which define the splice junction, the fourth column contains strand (1 or -1), the fifth column is the 
overhang (see definitions below), and the last column is the respective number of reads with these properties.

The secons output is a tab-delimited  file  which contains counts of continuous (non-split reads) which {\em overlap}
splice sites of splice junctions tabulated in the previous step. This second file is optional and is used to compute 
the completness of splicing index~\cite{pmid23172860, pmid22955974}.

\section{Definitions}

By definition, we say that we observe a splice junction each we see an 'N' symbol in the CIGAR attribute of some
SAM alignment. For instance, the alignment shown in Figure~\ref{fig::01} below gives rise to two splice junctions, 
denoted by $\SJ_1$ and $\SJ_2$.
%
\begin{figure}[h]
\footnotesize
\begin{verbatim}
       10        20        30        40        50        60        70        80
       |         |         |         |         |         |         |         |
       12345678 9012345678901234567890123456789012345678901234567890123456789012

Ref    AGTCTAGG*GACGGCATAGGAGGTGAGCATTTGTGTACGCAGATCTACAAAACATGTGTCACGGATAGGATCG
Query     CTAGGAGACGG**TAGGAG....................ATCTA*AAAACAT.............GATa
                            |<-----   SJ1  ----->|           |<--- SJ2 --->|
\end{verbatim}

The corresponding SAM line is:
\begin{verbatim}
Query   123   Ref   14    255    5M1I5M2D6M20N5M1D7M13N3M1S 1234 
\end{verbatim}
\caption{An example alignment and its CIGAR attribute\label{fig::01}}
\end{figure}
%
We keep the convention that coordinates of splice junctions always refer to terminal exonic nucleotides, 
i.e., $\SJ_1$ is $\rm Ref\_31\_52$ and $\SJ2$ is $\rm Ref\_64\_78$. We also denote the length of the 
intron by $l(\SJ)$, i.e. $l(\SJ_1)=52-31-1=20$ and $l(\SJ_2)=78-64-1=13$. Intron length is always equal 
to the corresponding 'N' number in the CIGAR attribute.

Each splice junctions is associated with four numbers: $m_u$ ($m_d$) --- the number of \underline{m}atching 
nucleotides immediately upstream (downstream) of the junction, and $v_u$ ($v_d$) --- the length in the 
reference of the aligned region, also called o\underline{v}erhang, which includes M/I/D CIGAR operations 
and is located immediately upstream (downstream) of the junction. In Figure~\ref{fig::01} we have 
$m_u(\SJ_1)=6$, $m_d(\SJ_1)=5$, $v_u(\SJ_1)=31-14+1=18$, $v_d(\SJ_1)=64-52+1=13$ and
$m_u(\SJ_2)=7$, $m_d(\SJ_2)=3$, $v_u(\SJ_2)=64-52+1=13$, $v_d(\SJ_2)=80-78+1=3$.

For each splice junction we require that
\begin{enumerate}
\item $l(\SJ)\ge \rm\bf minlen$ and $l(\SJ)\le \rm\bf maxlen$
\item $m_u\ge \rm\bf margin$ and $m_d\ge \rm\bf margin$
\end{enumerate}

In addition to junction coordinates, the overhang $v_u(\SJ)$ can also be used to correct for artifactually large read 
counts that arise in certain positions~\cite{pmid22537040}. In Figure~\ref{fig::02} we show six split reads supporting 
the same splice junction with overhangs 14 (Q1), 12 (Q2--Q4), and 8 (Q5--Q6).
%
\begin{figure}[h]
\footnotesize
\begin{verbatim}
       10       20        30        40        50        60        70        80
       |        |         |         |         |         |         |         |
       123456789012345678901234567890123456789012345678901234567890123456789012

Ref    AGTCTAGGGACGGCATAGGAGGTGAGCATTTGTGTACGCAGATCTACAAAACATGTGTCACGGATAGGATCG

Q1            GGACGGCATAGGAG....................ATCT      
Q2              ACGGCATAGGAG....................ATCTAC    
Q3              ACGGCATAGGAG....................ATCTAC    
Q4              ACGGCATAGGAG....................ATCTAC    
Q5                  CATAGGAG....................ATCTACAAAA
Q6                  CATAGGAG....................ATCTACAAAA
\end{verbatim}
\caption{Split reads support the same splice junction with different overhangs\label{fig::02}}
\end{figure}

The quantification of abundance is done as follows. For each splice junction (pair of coordinates) 
we initialize and keep $nbins$ separate counters. For each instance of a splice junction we increment 
the counter corresponding to the overhang bin defined by $d=floor(v_u/binsize)$.

For example, in the default settings we have $binsize=+\infty$. This means that $d=0$ for all supporting 
reads, regardless of their overhang ($v_u=14$ for Q1, $v_u=12$ for Q2--4, and $v_u=8$ for Q5--6 in 
Figure~\ref{fig::02}). Therefore, there is only one counter to increment, and the result will be the 
``collapsed'' counts. The output corresponding to Figure~\ref{fig::02} will then be
\begin{verbatim}
Ref     31      52      1       0       6
\end{verbatim}

By contrast, to take into account the overhang information, one should set $binsize=1$ (and also specify $nbins$ 
because the program doesn't know the range of possible overhang values). There will be a separate counter for each 
offset $d$ and the output corresponding to Figure~\ref{fig::02} will be
\begin{verbatim}
Ref     31      52      1       8        2
Ref     31      52      1       12       3
Ref     31      52      1       14       1
\end{verbatim}
Note that when aggregated by the fifth column, the number of counts coincides with the collapsed counts.
\bibliography{sjcount}
\bibliographystyle{abbrv}

\end{document}

